### DISCUSSION TITLE
Limit of 8 concurrent Background tasks in FAST API ?

### DISCUSSION THREAD

**Discussion by amardeep2006** [fastapi/fastapi#6474] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import time
from fastapi import BackgroundTasks, FastAPI
import uuid
app = FastAPI()


def write_notification(request_id):
    print("Sleeping for 5 seconds  "+str(request_id))
    time.sleep(5)
    print("rq id"+str(request_id))



@app.post("/send-notification")
async def send_notification(background_tasks: BackgroundTasks):
    
    request_id = uuid.uuid4()
    background_tasks.add_task(write_notification,request_id)
    return {"message": "Notification sent in the background"}
```

### Description
Hi All,
When I am hitting the send-notification end point from client, I am getting all the responses for my requests because the method is Async.

Server side behavior is not what I expect :
Expected behavior : All the background tasks must be executed as soon as they are added to BackgroundTasks.

Actual Behavior : After 8 requests , background tasks start piling up and wait for other tasks to complete. It appears something is restricting number of concurrent BackgroundTasks to eight.
Not sure if this is Starlette limitation.

I this something we can configure or override ?

Thanks in Advance.

Here are the logs 

INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  c0cb158a-f4d7-4371-aa08-ceec924d097a
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  55ae53c4-191f-42ba-b85d-0fd47f6d5a68
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  cccb7841-f0da-4f19-afee-b6bdaf12de29
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  967bb3e6-648b-445d-9afd-d393db74f1a0
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  2d709731-ed89-48e3-8f74-65393f9997ac
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  a28d7f7a-096f-41ff-b84e-1c8410048a4a
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  6ddc602b-05c5-4e5f-970f-143fb8d7b9c2
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  8f738fa6-9e8e-411a-94a4-0a9e57241844
**INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK**
rq idc0cb158a-f4d7-4371-aa08-ceec924d097a
Sleeping for 5 seconds  3973a355-2880-4f49-b838-bcbeb76d9c15
rq id55ae53c4-191f-42ba-b85d-0fd47f6d5a68
Sleeping for 5 seconds  c2391727-3dc0-44cc-a7e6-03eac3608406
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
rq idcccb7841-f0da-4f19-afee-b6bdaf12de29
Sleeping for 5 seconds  59c5aea9-b497-461e-9096-d92265452442
rq id967bb3e6-648b-445d-9afd-d393db74f1a0
Sleeping for 5 seconds  43d74ad7-0391-467a-ad63-09c55bd6586b
INFO:     127.0.0.1:52646 - "POST /send-notification HTTP/1.1" 200 OK
rq id2d709731-ed89-48e3-8f74-65393f9997ac
Sleeping for 5 seconds  8ecba683-2603-4b44-9170-196d8abc6604
rq ida28d7f7a-096f-41ff-b84e-1c8410048a4a
Sleeping for 5 seconds  bd72f365-1e6b-4d99-9b2c-fef144dd12e2
rq id6ddc602b-05c5-4e5f-970f-143fb8d7b9c2
Sleeping for 5 seconds  1e9b3d0c-6d3d-4e55-b7ab-cf7e5621006d
rq id8f738fa6-9e8e-411a-94a4-0a9e57241844
Sleeping for 5 seconds  434a5f80-e7fa-4a48-a3ec-af9a92a7076c
rq id3973a355-2880-4f49-b838-bcbeb76d9c15
Sleeping for 5 seconds  c84b6d55-97ef-4a26-b046-413712c4823f
rq idc2391727-3dc0-44cc-a7e6-03eac3608406
Sleeping for 5 seconds  2705a016-482d-4ee0-b09f-c116710498ba
rq id59c5aea9-b497-461e-9096-d92265452442
Sleeping for 5 seconds  d14ee180-424c-4075-80c2-351377b64c42
rq id43d74ad7-0391-467a-ad63-09c55bd6586b
Sleeping for 5 seconds  6245111e-dfe7-4411-a296-72edabfe2576
rq id8ecba683-2603-4b44-9170-196d8abc6604
Sleeping for 5 seconds  5c6bd2b8-af7c-4876-8357-7e37d1dbf901
rq idbd72f365-1e6b-4d99-9b2c-fef144dd12e2
Sleeping for 5 seconds  928a6167-da31-4281-a20d-c43adbc6b990
rq id1e9b3d0c-6d3d-4e55-b7ab-cf7e5621006d
Sleeping for 5 seconds  c6edd17d-c92f-4ef9-804f-f5f724bfec36
rq id434a5f80-e7fa-4a48-a3ec-af9a92a7076c
rq idc84b6d55-97ef-4a26-b046-413712c4823f
rq id2705a016-482d-4ee0-b09f-c116710498ba
rq idd14ee180-424c-4075-80c2-351377b64c42
rq id6245111e-dfe7-4411-a296-72edabfe2576
rq id5c6bd2b8-af7c-4876-8357-7e37d1dbf901
rq id928a6167-da31-4281-a20d-c43adbc6b990
rq idc6edd17d-c92f-4ef9-804f-f5f724bfec36

### Operating System

Windows

### Operating System Details

Windows 10 H2

### FastAPI Version

Version: 0.68.1

### Python Version

3.8.10

**Comment by wcastello** [comment]
How are you doing those requests and measuring that they pile after 8 @amardeep2006? Also, did you pass any parameters to uvicorn? 

I added a datetime.now() to your write_notification() and ran a

```
for i in {1..16}
do
curl -X POST 127.0.0.1:8000/send-notification
done
```

It seems to work as you'd expect: 
```
INFO:     127.0.0.1:50611 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  177ee6db-87c6-445c-ab52-a58fbd62ded12021-12-18 14:31:15.593910
INFO:     127.0.0.1:50612 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  e6c5dc5e-8fe0-4118-af0d-b277d601229f2021-12-18 14:31:15.607869
INFO:     127.0.0.1:50613 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  a5efa0dc-af87-4f5e-8303-7bf73c4976672021-12-18 14:31:15.620553
INFO:     127.0.0.1:50614 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  2b64ffa0-1ae8-45d3-9fa5-41d13e9d75862021-12-18 14:31:15.633375
INFO:     127.0.0.1:50615 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  6d91f741-4f0e-49d9-9f06-102915103aff2021-12-18 14:31:15.645207
INFO:     127.0.0.1:50616 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  a300b2ae-57d7-4bfc-a040-a59fb532af6e2021-12-18 14:31:15.657757
INFO:     127.0.0.1:50617 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  b44f4a15-688f-4d4b-8d4d-643027e9474d2021-12-18 14:31:15.670397
INFO:     127.0.0.1:50618 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  7d846633-0d85-4fae-b2fb-20e9022f87e72021-12-18 14:31:15.683769
INFO:     127.0.0.1:50619 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  9263dce5-e500-4d75-b339-6d2a6d58a4442021-12-18 14:31:15.696140
INFO:     127.0.0.1:50620 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  01e6507d-8a84-483c-9b6b-5cd075c890682021-12-18 14:31:15.708734
INFO:     127.0.0.1:50621 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  969a2f05-3f9f-4ee6-9a1e-f14902e5628a2021-12-18 14:31:15.724446
INFO:     127.0.0.1:50622 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  8400658c-c9ad-4139-8e64-d868f55cd1772021-12-18 14:31:15.737273
INFO:     127.0.0.1:50623 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  7a4d4cf0-48e4-4a16-a855-d900fadf5f052021-12-18 14:31:15.749940
INFO:     127.0.0.1:50624 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  ad706dec-a56c-4316-9c98-9d92777c7bf82021-12-18 14:31:15.762147
INFO:     127.0.0.1:50625 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  552bec72-5c7f-416b-a1bb-cb2f07f9e3872021-12-18 14:31:15.775507
INFO:     127.0.0.1:50626 - "POST /send-notification HTTP/1.1" 200 OK
Sleeping for 5 seconds  1b18653a-28ec-42f0-aec5-7ecc910af7812021-12-18 14:31:15.788307
rq id177ee6db-87c6-445c-ab52-a58fbd62ded12021-12-18 14:31:20.597017
rq ide6c5dc5e-8fe0-4118-af0d-b277d601229f2021-12-18 14:31:20.608636
rq ida5efa0dc-af87-4f5e-8303-7bf73c4976672021-12-18 14:31:20.624223
rq id2b64ffa0-1ae8-45d3-9fa5-41d13e9d75862021-12-18 14:31:20.635446
rq id6d91f741-4f0e-49d9-9f06-102915103aff2021-12-18 14:31:20.647588
rq ida300b2ae-57d7-4bfc-a040-a59fb532af6e2021-12-18 14:31:20.660428
rq idb44f4a15-688f-4d4b-8d4d-643027e9474d2021-12-18 14:31:20.673079
rq id7d846633-0d85-4fae-b2fb-20e9022f87e72021-12-18 14:31:20.686956
rq id9263dce5-e500-4d75-b339-6d2a6d58a4442021-12-18 14:31:20.699033
rq id01e6507d-8a84-483c-9b6b-5cd075c890682021-12-18 14:31:20.710383
rq id969a2f05-3f9f-4ee6-9a1e-f14902e5628a2021-12-18 14:31:20.728717
rq id8400658c-c9ad-4139-8e64-d868f55cd1772021-12-18 14:31:20.741334
rq id7a4d4cf0-48e4-4a16-a855-d900fadf5f052021-12-18 14:31:20.753921
rq idad706dec-a56c-4316-9c98-9d92777c7bf82021-12-18 14:31:20.763207
rq id552bec72-5c7f-416b-a1bb-cb2f07f9e3872021-12-18 14:31:20.777005
rq id1b18653a-28ec-42f0-aec5-7ecc910af7812021-12-18 14:31:20.792848
```

**Comment by raphaelauv** [comment]
https://github.com/tiangolo/fastapi/issues/4221

https://github.com/tiangolo/fastapi/issues/4107#issuecomment-974797385

**Comment by amardeep2006** [comment]
> How are you doing those requests and measuring that they pile after 8 @amardeep2006? Also, did you pass any parameters to uvicorn?
> 
> I added a datetime.now() to your write_notification() and ran a
> 
> ```
> for i in {1..16}
> do
> curl -X POST 127.0.0.1:8000/send-notification
> done
> ```
> 
> It seems to work as you'd expect:
> 
> ```
> INFO:     127.0.0.1:50611 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  177ee6db-87c6-445c-ab52-a58fbd62ded12021-12-18 14:31:15.593910
> INFO:     127.0.0.1:50612 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  e6c5dc5e-8fe0-4118-af0d-b277d601229f2021-12-18 14:31:15.607869
> INFO:     127.0.0.1:50613 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  a5efa0dc-af87-4f5e-8303-7bf73c4976672021-12-18 14:31:15.620553
> INFO:     127.0.0.1:50614 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  2b64ffa0-1ae8-45d3-9fa5-41d13e9d75862021-12-18 14:31:15.633375
> INFO:     127.0.0.1:50615 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  6d91f741-4f0e-49d9-9f06-102915103aff2021-12-18 14:31:15.645207
> INFO:     127.0.0.1:50616 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  a300b2ae-57d7-4bfc-a040-a59fb532af6e2021-12-18 14:31:15.657757
> INFO:     127.0.0.1:50617 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  b44f4a15-688f-4d4b-8d4d-643027e9474d2021-12-18 14:31:15.670397
> INFO:     127.0.0.1:50618 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  7d846633-0d85-4fae-b2fb-20e9022f87e72021-12-18 14:31:15.683769
> INFO:     127.0.0.1:50619 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  9263dce5-e500-4d75-b339-6d2a6d58a4442021-12-18 14:31:15.696140
> INFO:     127.0.0.1:50620 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  01e6507d-8a84-483c-9b6b-5cd075c890682021-12-18 14:31:15.708734
> INFO:     127.0.0.1:50621 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  969a2f05-3f9f-4ee6-9a1e-f14902e5628a2021-12-18 14:31:15.724446
> INFO:     127.0.0.1:50622 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  8400658c-c9ad-4139-8e64-d868f55cd1772021-12-18 14:31:15.737273
> INFO:     127.0.0.1:50623 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  7a4d4cf0-48e4-4a16-a855-d900fadf5f052021-12-18 14:31:15.749940
> INFO:     127.0.0.1:50624 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  ad706dec-a56c-4316-9c98-9d92777c7bf82021-12-18 14:31:15.762147
> INFO:     127.0.0.1:50625 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  552bec72-5c7f-416b-a1bb-cb2f07f9e3872021-12-18 14:31:15.775507
> INFO:     127.0.0.1:50626 - "POST /send-notification HTTP/1.1" 200 OK
> Sleeping for 5 seconds  1b18653a-28ec-42f0-aec5-7ecc910af7812021-12-18 14:31:15.788307
> rq id177ee6db-87c6-445c-ab52-a58fbd62ded12021-12-18 14:31:20.597017
> rq ide6c5dc5e-8fe0-4118-af0d-b277d601229f2021-12-18 14:31:20.608636
> rq ida5efa0dc-af87-4f5e-8303-7bf73c4976672021-12-18 14:31:20.624223
> rq id2b64ffa0-1ae8-45d3-9fa5-41d13e9d75862021-12-18 14:31:20.635446
> rq id6d91f741-4f0e-49d9-9f06-102915103aff2021-12-18 14:31:20.647588
> rq ida300b2ae-57d7-4bfc-a040-a59fb532af6e2021-12-18 14:31:20.660428
> rq idb44f4a15-688f-4d4b-8d4d-643027e9474d2021-12-18 14:31:20.673079
> rq id7d846633-0d85-4fae-b2fb-20e9022f87e72021-12-18 14:31:20.686956
> rq id9263dce5-e500-4d75-b339-6d2a6d58a4442021-12-18 14:31:20.699033
> rq id01e6507d-8a84-483c-9b6b-5cd075c890682021-12-18 14:31:20.710383
> rq id969a2f05-3f9f-4ee6-9a1e-f14902e5628a2021-12-18 14:31:20.728717
> rq id8400658c-c9ad-4139-8e64-d868f55cd1772021-12-18 14:31:20.741334
> rq id7a4d4cf0-48e4-4a16-a855-d900fadf5f052021-12-18 14:31:20.753921
> rq idad706dec-a56c-4316-9c98-9d92777c7bf82021-12-18 14:31:20.763207
> rq id552bec72-5c7f-416b-a1bb-cb2f07f9e3872021-12-18 14:31:20.777005
> rq id1b18653a-28ec-42f0-aec5-7ecc910af7812021-12-18 14:31:20.792848
> ```

Hi @wcastello : I have done some more analysis on this issue it seems the limit is based on available CPU on the FAST API server. I am running it on a 2 core processor with 4 logical Processor .

![image](https://user-images.githubusercontent.com/50517840/146717767-e3d6e91a-82fd-42f4-8932-3d6b3f077763.png)

Can you please share number of CPU cores on your machine ?

Regarding your another question : I am using this pattern in actual application where background tasks are suppose to update DB continuously , which stops happening after 8 requests. Also in this sample program , You can see that logs show that BG tasks does not start immediately .
There is no other parameter passed to uvicorn. I am starting app by using following command.
`uvicorn app:app` 

Here is another comment which talks about CPU dependency on number of threads limit. May be this issue is related to CPU limitations in my case as well.

https://github.com/tiangolo/fastapi/issues/603#issuecomment-540834120

Summary of that comment : 
> If you look at the code for starlette's run_in_threadpool function, it is just using the built-in python ThreadPoolExecutor (with default settings) to run the function. If you look in the standard library, you'll see max_workers = (os.cpu_count() or 1) * 5, so the maximum number of threads per process should be bounded by that value.

**Comment by amardeep2006** [comment]
I am closing this issue as I am not facing this on server which has higher capacity in term of CPU.
